File: compiler/src/tree_parser.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterator
from io import StringIO
from pathlib import Path
from strutil import cut, extract_indent
from node import Node, Position
from typing import Generic, List, TypeVar
from logger import default_logger

T = TypeVar("T")

class TrimStack(Generic[T]):
    def __init__(self) -> None:
        self._data: List[T] = []

    def __setitem__(self, index: int, value: T) -> None:
        if index == len(self._data):
            self._data.append(value)
        elif 0 <= index < len(self._data):
            self._data[index] = value
        else:
            raise IndexError("Index out of bounds (can only set existing index or append)")

    def __getitem__(self, index: int) -> T:
        return self._data[index]

    def trim(self, max_index: int) -> None:
        if max_index < 0:
            self._data.clear()
        else:
            self._data = self._data[:max_index + 1]

    def __len__(self) -> int:
        return len(self._data)

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({self._data})"

class TreeParser:    
    def parse_tree(self, code: str) -> Node:
        # prepend and append newlines for simpler and cleaner handling
        # TODO - although this will fuck over line numbers, so might actually be a bad idea?
        code = f"\n{code}\n"
        
        default_logger.log("parse", f"parsing {len(code)} characters of source code")

        @dataclass
        class ParsingNode:
            content: str
            position: Position
            children: list[ParsingNode] = field(default_factory=list)            
            def toNode(self):
                return Node(content=self.content, pos=self.position, children=[child.toNode() for child in self.children])

        scope: TrimStack[ParsingNode] = TrimStack()
        root = ParsingNode("PIL:file", Position(0))
        scope[0] = root
        line_num = 0
        
        lines = code.split("\n")
        default_logger.log("parse", f"processing {len(lines)} lines")
        
        for line in lines:
            line_num += 1 # at the start - assumes above \n{code}\n
            line, indent = extract_indent(line)

            # simplifies code. all the top-level lines are indent-1, belonging to a fake top-level Node
            # which is at indent-0
            indent += 1

            if line.isspace() or len(line) <= 0:
                # skip empty/indentation-only lines
                continue

            default_logger.log("parse", f"line {line_num}: indent={indent}, content='{line}'")
            node = ParsingNode(line, Position(line_num))
            scope[indent - 1].children.append(node)
            scope[indent] = node
            scope.trim(indent)
        
        result = root.toNode()
        default_logger.log("parse", f"parsing complete, tree has {self._count_nodes(result)} nodes")
        return result
        
    def _count_nodes(self, node: Node) -> int:
        """helper to count total nodes in tree for logging"""
        count = 1
        for child in node.children:
            count += self._count_nodes(child)
        return count
        

File: compiler/src_ts/core/tree_parser.ts
// core/tree_parser.ts

import { extractIndent } from "../utils/strutil.ts";
import { Node, Position } from "./node.ts";
import { default_logger, LOGGER_NO_CONTEXT } from "../utils/logger.ts";

class Trim_stack<T> {
  private data: T[] = [];

  set(index: number, value: T): void {
    if (index === this.data.length) {
      this.data.push(value);
    } else if (index >= 0 && index < this.data.length) {
      this.data[index] = value;
    } else {
      throw new RangeError(
        "Index out of bounds (can only set existing index or append)",
      );
    }
  }

  get(index: number): T | undefined {
    return this.data[index];
  }

  trim(max_index: number): void {
    if (max_index < 0) {
      this.data = [];
    } else {
      const new_size = Math.min(max_index + 1, this.data.length);
      this.data = this.data.slice(0, new_size);
    }
  }

  get size(): number {
    return this.data.length;
  }

  toString(): string {
    return `${this.constructor.name}(${JSON.stringify(this.data)})`;
  }
}

class Parsing_node {
  readonly content: string;
  readonly position: Position;
  readonly children: Parsing_node[];

  constructor(
    content: string,
    position: Position,
    children: Parsing_node[] = [],
  ) {
    this.content = content;
    this.position = position;
    this.children = children;
  }

  to_Node(): Node {
    return new Node(
      this.content,
      this.position,
      this.children.map((c) => c.to_Node()),
    );
  }
}

export class Tree_parser {
  private top_level: string;
  constructor(opts: {
    top_level: string
  }) {
    this.top_level = opts.top_level;
  }
  parse_tree(code: string, _compiler: unknown = null): Node {
    if (code.endsWith("\n")) {
      code = code.slice(0, -1);
    }

    // wrap with newlines like Crystal
    code = `\n${code}\n`;

    const scope = new Trim_stack<Parsing_node>();
    const root = new Parsing_node(this.top_level, new Position(0));
    scope.set(0, root);
    let line_num = 0;

    const lines = code.split("\n");

    for (const raw of lines) {
      line_num += 1;

      let [line, extracted_indent] = extractIndent(raw, scope.size - 1);
      let indent = extracted_indent + 1;

      if (line.trim().length === 0) {
        indent = scope.size;
      }

      line = line.trimEnd();

      default_logger.log(
        LOGGER_NO_CONTEXT, // but TODO contextualize by line
        "parse",
        `line ${line_num}: indent=${indent}, content='${line}'`,
      );

      const parent = scope.get(indent - 1);
      if (parent === undefined) {
        throw new Error(
          `Invalid indentation at line ${line_num}: ${raw}`,
        );
      }
      const node = new Parsing_node(line, new Position(line_num));
      parent.children.push(node);
      scope.set(indent, node);
      scope.trim(indent);
    }

    return root.to_Node();
  }
}
